src/filters/list.js

```diff
@@ -0,0 +1,197 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module paste-from-office/filters/list
+ */
+
+import Element from '@ckeditor/ckeditor5-engine/src/view/element';
+import Matcher from '@ckeditor/ckeditor5-engine/src/view/matcher';
+import Position from '@ckeditor/ckeditor5-engine/src/view/position';
+import TreeWalker from '@ckeditor/ckeditor5-engine/src/view/treewalker';
+import UpcastWriter from '@ckeditor/ckeditor5-engine/src/view/upcastwriter';
+
+/**
+ * Transforms Word specific list-like elements to the semantic HTML lists.
+ *
+ * Lists in Word are represented by block elements with special attributes like:
+ *
+ *		<p class=MsoListParagraphCxSpFirst style='mso-list:l1 level1 lfo1'>...</p> // Paragraph based list.
+ *		<h1 style='mso-list:l0 level1 lfo1'>...</h1> // Heading 1 based list.
+ *
+ * @param {module:engine/view/node~Node|module:engine/view/documentfragment~DocumentFragment} bodyView The view
+ * structure which to transform.
+ * @returns {module:engine/view/node~Node|module:engine/view/documentfragment~DocumentFragment} The view
+ * structure instance with list-like elements transformed into semantic lists.
+ */
+export function transformParagraphsToLists( bodyView, stylesString ) {
```

Comment:
So, how I read this function is:

1. Get first child of body view... Why of a body? Can't this be of any container?
2. Then... if the first child exists do something. Why is this first child that important? I thought we're filtering all children.
3. Now, we want to find all list nodes... at a position before the first child?
4. And we create lists for these... list nodes? Where do we create them? If we create them why isn't `createLists()` returning them?
5. And then we return the `bodyView` again. Which can be a `Node`? Or a doc frag? How can a body be both. This function is used in one place in the code, so why do we support both things?

Reply:
How to improve this code? You need to think about signatures of the functions that you create. And how you use them and combine them together.

1. `transformParagraphsToLists()` should support just one thing – either a container or a doc frag. Not both. You can choose what you need based on the most common use of this function. And there's just one place where you use it so it'll be easy :P.
2. If you want to keep the outline of this function similar to what it looks now then `findAllListNodes()` should be corrected as well. It's unjustified from the signature of this function to work on a position. In 99.9% of cases you look for stuff in a container or a range. Even `TreeWalker` does that. It can accept a start position but this isn't necessary to pass it there (in this case where you scan the whole body anyway). Nor should this fact leak and pollute an unrealted method like `transformParagraphsToLists()`.
3. Then, once you'll have all list nodes... Wait... are these list nodes already? I thought they are paragraphs (based on the name of this outer function). So they are neither lists nor nodes. They are list-item-like-elements. The naming should be consequent in your code. Furthermore, from the docs I saw that `<h1>` may represent a list item as well so... `transformParagraphsToLists()` should also be renamed.
4. OK, so we finally have that `listItemLikeElements` array. Let's transform these elements now. But not via a `createLists()` function unless we're creating lists. We are transforming X into Y. Again – naming needs to be adjusted.
5. Finally, to balance the amount of logic between functions, I'd recommend moving the loop out of the current `createLists()` to the outer function because it'll make that singular `transformXIntoY()` easier to test and debug.
6. Returning `bodyView` isn't necessary at this stage. I know that you anticipate there to be filters which may return something else than they accepted, but I can't see such a filter today so KISS. Don't build abstractions/mechanisms which you don't need yet.